### PR TITLE
[dev] Migrate full charm and unit agent state

### PR DESCRIPTION
**Pull request by achilleasa**
## Description of change

This PR ensures that each of the following pieces of server-side persisted unit agent data are correctly exported/imported when migrating models across controllers:
- charm state
- uniter worker internal state
- uniter worker storage state
- uniter worker relation data state (storage of the actual relation data is in progress in a separate PR)
- meter status worker internal state.

## QA steps (run on **AWS**)

```console
$ juju bootstrap aws mig-dst --no-gui
$ juju bootstrap aws mig-src --no-gui

$ juju whoami
Controller:  mig-src
Model:       default
User:        admin

# Deploy some charms and populate the various bits of state data (except relation data)
# stored server-side.
$ juju deploy ./testcharms/charm-repo/bionic/metered-multi-series --series bionic
$ juju deploy swift-storage

# Wait for things to settle down; then:
$ juju run --unit metered-multi-series/0 state-set foo=bar
$ juju run --unit swift-storage/0 storage-add block-devices
$ juju list-storage
Unit             Storage id       Type   Pool  Size    Status    Message
swift-storage/0  block-devices/0  block  loop  1.0GiB  attached

# Migrate the model
$ juju destroy-model -y mig-dst:default
$ juju migrate default mig-dst
$ juju switch mig-dst:default

# Attach to mongo and verify migrated unit states
juju:PRIMARY> db.unitstates.find().pretty()
{
        "_id" : "194c0589-5065-4ca9-8f53-6d2b7b7de5a8:u#metered-multi-series/0#charm",
        "txn-revno" : NumberLong(9),
        "charm-state" : {
                "foo" : "bar"
        },
        "uniter-state" : "leader: true\nstarted: true\nstopped: false\ninstalled: true\nremoved: false\nstatus-set: true\nop: continue\nopstep: pending\nconfig-hash: 98d9c7b7c9d4d36d16ab25ac4d1c5772e47c7c185bb0ab66b98ce6e3990e9b7e\ntrust-hash: d0d561b092b8d50dd3d9af63e87679862e4a6e558b728ac509a79c2e59ea9166\naddresses-hash: 41dde5360171cce9036d1825bb1908439bfc8e2eac5f93b1fe5b72c170a9b984\n",
        "meter-status-state" : "status-code: AMBER\nstatus-info: not set\n",
        "model-uuid" : "194c0589-5065-4ca9-8f53-6d2b7b7de5a8",
        "txn-queue" : [
                "5e8c73c0d9ad0e11d559ee38_f99573ec"
        ]
}
{
        "_id" : "194c0589-5065-4ca9-8f53-6d2b7b7de5a8:u#swift-storage/0#charm",
        "txn-revno" : NumberLong(8),
        "uniter-state" : "leader: true\nstarted: true\nstopped: false\ninstalled: true\nremoved: false\nstatus-set: true\nop: continue\nopstep: pending\nconfig-hash: 7abf88e3e462f6b309e718edd554f038e75994286f3ba77549ce579de9984e20\ntrust-hash: 5f031a6a1b38dbd4f201632b0d8b2a2f37236b4c759a4de1a48046e7d20adc55\naddresses-hash: fc4deb1f962b4451b1e2d548fec03d823a69ac2a68f1c4f445126fb7aae4a2ff\n",
        "storage-state" : "block-devices/0: true\n",
        "meter-status-state" : "status-code: AMBER\nstatus-info: not set\n",
        "model-uuid" : "194c0589-5065-4ca9-8f53-6d2b7b7de5a8",
        "txn-queue" : [
                "5e8c73cad9ad0e11d559eef1_53174b78"
        ]
}
```

### CODE REVIEW

**Comment by achilleasa** [comment]
!!build!! (tests did not start)

File: state/migration_import.go [hunk 1]
```diff
@@ -1174,6 +1164,36 @@ func (i *importer) unit(s description.Application, u description.Unit, ctrlCfg c
...
+	}
+	if storageState := u.StorageState(); storageState != "" {
+		us.SetStorageState(storageState)
+	}
+	if meterStatusState := u.MeterStatusState(); meterStatusState != "" {
+		us.SetMeterStatusState(meterStatusState)
+	}
+
+	// No state to persist.
+	if !us.Modified() {
```
Comment (SimonRichardson) [inline state/migration_import.go:1187]: Can we test this scenario?

Comment (achilleasa) [reply state/migration_import.go:1187]: What would be the purpose of this test? That we don't call SetState if the description is empty or that `us.Modified()` works as expected (the latter is already tested in state)

Comment (SimonRichardson) [reply state/migration_import.go:1187]: I want to see that we don't call SetState :)

**Comment by achilleasa** [comment]
$$merge$$

**Pull request merged**